modules/assim.sequential/R/downscale_function.R

```diff
@@ -1,88 +1,284 @@
+
+SDA_downscale <- function(preprocessed, date, carbon_pool, covariates, model_type) {
+  input_data <- preprocessed$input_data
+  site_coordinates <- preprocessed$site_coordinates
+  carbon_data <- preprocessed$carbon_data
+  
+  # Convert site coordinates to SpatVector
+  site_coordinates <- terra::vect(site_coordinates, geom = c("lon", "lat"), crs = "EPSG:4326")
+  
+  # Extract predictors from covariates raster using site coordinates
+  predictors <- as.data.frame(terra::extract(covariates, site_coordinates, ID = FALSE))
+  
+  # Dynamically get covariate names
+  covariate_names <- names(predictors)
+  
+  # Create a single data frame with all predictors and ensemble data
+  full_data <- cbind(carbon_data, predictors)
+  
+  # Split the observations into training and testing sets
+  set.seed(123)  # for reproducibility
+  sample <- sample(1:nrow(full_data), size = round(0.75 * nrow(full_data)))
+  train_data <- full_data[sample, ]
+  test_data <- full_data[-sample, ]
+  
+  if (model_type == "rf") {
+    # Train a random forest model for each ensemble member using the training data
+    rf_output <- list()
+    for (i in seq_along(carbon_data)) {
+      ensemble_col <- paste0("ensemble", i)
+      formula <- as.formula(paste(ensemble_col, "~", paste(covariate_names, collapse = " + ")))
+      rf_output[[i]] <- randomForest::randomForest(formula,
+                                                   data = train_data,
+                                                   ntree = 1000,
+                                                   na.action = stats::na.omit,
+                                                   keep.forest = TRUE,
+                                                   importance = TRUE)
+    }
+    
+    # Generate predictions (maps) for each ensemble member using the trained models
+    maps <- list()
+    predictions <- list()
+    for (i in seq_along(rf_output)) {
+      maps[[i]] <- terra::predict(covariates, model = rf_output[[i]], na.rm = TRUE)
+      predictions[[i]] <- predict(rf_output[[i]], test_data)
+    }
+    
+    # Organize the results into a single output list
+    downscale_output <- list(
+      data = list(training = train_data, testing = test_data),
+      models = rf_output,
+      maps = maps,
+      predictions = predictions
+    )
+    
+  } else if (model_type == "cnn") {
+    # Prepare data for CNN
+    x_data <- as.matrix(full_data[, covariate_names])
+    y_data <- as.matrix(carbon_data)
+    
+    # Calculate scaling parameters from all data
+    scaling_params <- list(
+      mean = colMeans(x_data),
+      sd = apply(x_data, 2, sd)
+    )
+    
+    # Normalize the data
+    x_data_scaled <- scale(x_data, center = scaling_params$mean, scale = scaling_params$sd)
+    
+    # Split into training and testing sets
+    x_train <- x_data_scaled[sample, ]
+    x_test <- x_data_scaled[-sample, ]
+    y_train <- y_data[sample, ]
+    y_test <- y_data[-sample, ]
```

Comment:
move date prep tasks that are (or should be) identical across ML options to be before the `if(model_type...`

Reply:
addressed in commit [bb66142](https://github.com/PecanProject/pecan/pull/3321/commits/bb6614200471192b697c84a1ed87be8720272f9b)